pkg/traces/promsdprocessor/consumer/consumer.go

```diff
@@ -0,0 +1,249 @@
+	"github.com/prometheus/common/model"
+	"go.opentelemetry.io/collector/client"
+	otelcomponent "go.opentelemetry.io/collector/component"
+	otelconsumer "go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	semconv "go.opentelemetry.io/collector/semconv/v1.5.0"
+)
+
+const (
+	// OperationTypeInsert inserts a new k/v if it isn't already present
+	OperationTypeInsert = "insert"
+	// OperationTypeUpdate only modifies an existing k/v
+	OperationTypeUpdate = "update"
+	// OperationTypeUpsert does both of above
+	OperationTypeUpsert = "upsert"
+
+	//TODO: It'd be cleaner to get these from the otel semver package?
+	//      Not all are in semver though. E.g. "k8s.pod.ip" is internal inside the k8sattributesprocessor.
+	PodAssociationIPLabel       = "ip"
+	PodAssociationOTelIPLabel   = "net.host.ip"
+	PodAssociationk8sIPLabel    = "k8s.pod.ip"
+	PodAssociationHostnameLabel = "hostname"
+	PodAssociationConnectionIP  = "connection"
+)
+
+func ValidateOperationType(operationType string) error {
+	switch operationType {
+	case
+		OperationTypeUpsert,
+		OperationTypeInsert,
+		OperationTypeUpdate:
+		// Valid configuration, do nothing.
+	default:
+		return fmt.Errorf("unknown operation type %s", operationType)
+	}
+	return nil
+}
+
+func ValidatePodAssociations(podAssociations []string) error {
+	for _, podAssociation := range podAssociations {
+		switch podAssociation {
+		case
+			PodAssociationIPLabel,
+			PodAssociationOTelIPLabel,
+			PodAssociationk8sIPLabel,
+			PodAssociationHostnameLabel,
+			PodAssociationConnectionIP:
+			// Valid configuration, do nothing.
+		default:
+			return fmt.Errorf("unknown pod association %s", podAssociation)
+		}
+	}
+	return nil
+}
+
+// TODO: Put a private member so that this can't be created without calling NewConsumer?
+type Consumer struct {
+	nextConsumer otelconsumer.Traces
+
+	mtx        sync.Mutex
+	hostLabels map[string]discovery.Target
+
+	operationType   string
+	podAssociations []string
+
+	logger log.Logger
+}
+
+var _ otelconsumer.Traces = (*Consumer)(nil)
+
+func NewConsumer(nextConsumer otelconsumer.Traces, operationType string, podAssociations []string, logger log.Logger) (*Consumer, error) {
+	if nextConsumer == nil {
+		return nil, otelcomponent.ErrNilNextConsumer
+	}
+
+	err := ValidateOperationType(operationType)
+	if err != nil {
+		return nil, err
+	}
+
+	err = ValidatePodAssociations(podAssociations)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Consumer{
+		nextConsumer:    nextConsumer,
+		hostLabels:      make(map[string]discovery.Target),
+		operationType:   operationType,
+		podAssociations: podAssociations,
+		logger:          logger,
+	}, nil
+}
+
+func (c *Consumer) SetHostLabels(hostLabels map[string]discovery.Target) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	c.hostLabels = hostLabels
+}
+
+func (c *Consumer) GetHostLabels(ip string) (map[string]string, error) {
+	c.mtx.Lock()
```

Comment:
nit: RLock/RUnlock instead? Would need to change to RWMutex

Reply:
Actually, I just realised there is a bigger issue 😬  This function returns a map which points to the same values. The map is not deep copied. The original code [doesn't have this issue](https://github.com/grafana/agent/blob/v0.35.1/pkg/traces/promsdprocessor/prom_sd_processor.go#L174-L177) because it locks a mutex for as long as we need to read the values in the map.

I deleted this function and switched back to the original implementation's choice of locking a mutex for as long as we need to read the data.

Thank you for bringing this to attention!

Reply:
Is this all resolved now?

Reply:
Yes. Now we just lock the mutex for [as long as the data is being read](https://github.com/grafana/agent/blob/2299-flow-otelcolprocessordiscovery-component/pkg/traces/promsdprocessor/consumer/consumer.go#L153-L175).